Add 'queue-search' command. Only send commands to single tab. Check user interaction before attempting hotkey playback.

frypf opened the pull request:
Hi @zopieux, really appreciate your efforts with this extension over the last few years - especially when Spotify keep changing their classes/selectors etc. or ab-testing new svgs!

Further to [my comment in #14 ](https://github.com/zopieux/spotify-hotkeys/issues/14#issuecomment-1675054263) I thought I'd have a go at determining the best single Spotify tab to issue with commands.

During the process I thought it might also be useful to add a global hotkey for switching to (or creating) a Spotify tab, and then cycling through the home / queue / search pages.

Unfortunately this highlighted another issue where the Spotify page itself might throw `DOMException: play() failed because the user didn't interact with the document first`.

I've actually had this happen a few times even before making my modifications, but I guess having a hotkey to create a Spotify tab directly just made it more obvious what was going on. The pause button would correctly switch to the play icon but with no audio output. The only way to get the audio back in this situation seems to be playing a different song or playlist directly or skipping tracks (via either a mouse click or Spotify's native in-page shortcut handlers).

Anyway I've tried mitigating this by force-focusing the tab and presenting a dialog/modal to highlight the issue to the user, and then registering self-removing listeners to correctly dispatch the command only after a keypress or mouse click anywhere within the page.

This ended up being a large chunk of extra code for a fairly niche situation, so I split it into a separate script (mostly just for easier comprehension for myself, tbh). Hopefully my modifications are useful, apologies if anything isn't clear or could use some refinement - I appreciate the way your code cascades intuitively and obviously, and doesn't rely on any chrome messaging. But either way thanks again for the extension itself 👍.

zopieux commented on src/web_accessible/checkPageInteraction.js:
```diff
@@ -0,0 +1,107 @@
+				if (!window.navigator.userActivation.hasBeenActive) return;
+				['keydown', 'pointerdown'].forEach((evType) => window.removeEventListener(evType, reactToUserInteraction));
+				document.querySelectorAll('[id*="swph__"').forEach((e) => e.remove());
+				window.lastCommand && successCallback(window.lastCommand);
```
Is there any way we can avoid polluting the "window" namespace? Wouldn't an root-scoped variable work, or maybe the native state API?

frypf replied:
I was trying to come up with a way around this and felt irritated leaving it like this in the first place. But it's not straightforward as all the code is effectively working as an IIFE due to being injected via the 
```javascript
(await import(chrome.runtime.getURL("web_accessible/checkPageInteraction.js"))).default(command, dispatchCommand);
```
line. Unfortunately the chrome storage api can apparently only be used from content scripts if you call
```javascript
chrome.storage.session.setAccessLevel({ accessLevel: 'TRUSTED_AND_UNTRUSTED_CONTEXTS' });
```
first from within the service worker, and while I'm _relatively_ sure any security risks would be minimal, it still didn't sit right with me doing this either.

Anyway I completely overlooked just using a dataset prop on the injected dialog itself, so I reckon that might be the way to go.

zopieux reviewed (changes requested):
Thanks for the kind words and thanks for taking the time to write a non-trivial PR. I left a lot of relatively minor comments, I'd appreciate if you could address them, keeping the changes in their respective commits for atomicity.

The PR otherwise looks good to me and will be merged. :+1:

zopieux commented on src/background.js:
```diff
@@ -1,20 +1,20 @@
           setTimeout(() => anim(i - 1), 80);
         };
-        anim(3);
+        anim(2);
```
I've tested your code and while this looks nice on buttons, the animation on the sidebar navigation links looks janky. Please add a condition to only animate when tag == button (or pass a bool) in `clickAndAnimate`.

frypf replied:
Agreed.

frypf commented on src/web_accessible/checkPageInteraction.js:
```diff
@@ -0,0 +1,107 @@
+		// Check user interacted with page before executing commands.
+		!window.navigator.userActivation.hasBeenActive &&
+		// Skip check if audio routed to other device (or tab).
+		!document.querySelector('#device-picker-icon-button').classList.contains('control-button--active')
```
None, beyond my need to pointlessly over-complicate things 😅.

frypf commented on src/background.js:
```diff
@@ -1,20 +1,20 @@
-  // classes... but not for all UI elements of course. Yay, consistency!
-  function findAndClick(command) {
-    // https://github.com/mantou132/Spotify-Lyrics/issues/94
+  // Let's have multiple methods to find the right button, because Spotify
```
Apologies for this - I have my IDE set up to highlight comments followed by an emspace in a different colour, which I use for intended info-type comments like this so that I can visually scan whether I've left any commented-out test code in a file. I apparently just staged the whole file on that commit rather than chunking changes, sorry.

zopieux replied:
No worries!

frypf commented on src/_locales/en/messages.json:
```diff
@@ -39,5 +42,36 @@
+                "example": "inserted via `chrome.runtime.getURL('web_accessible/logo.svg')` within js"
+            },
+            "TITLE": {
+                "content": "Spotify Web Player Hotkeys",
```
I'll rework the dialog HTML as per your suggestions below and probably just refactor this whole mess into 2 simple strings for dialog `text` vs `subtext`.

frypf commented on src/_locales/en/messages.json:
```diff
@@ -39,5 +42,36 @@
+        "message": "<p><img src=\"$LOGO$\"><span><strong>$TITLE$</strong></p>\n<p>$MESSAGE$ $CMD_TXT$</span></p>\n<p>• $INSTRUCTION$ <em>($CAVEAT$ <strong>esc</strong>)</em></p>",
+        "description": "presented to user when hotkey playback attempted before page interaction",
+        "placeholders": {
+            "LOGO": {
```
This was just to make it obvious what was going where, but I doubt there'll be any need for substitutions anyway once the dialog is simplified.

zopieux replied:
Indeed, ACK.

frypf commented on src/_locales/en/messages.json:
```diff
@@ -39,5 +42,36 @@
     "volume_mute": {
         "message": "Volume: toggle mute"
+    },
+    "interact_dialog_text": {
```
Totally agree re: the clutter, I was just trying to make it obvious why some commands are still allowed to execute while `play-pause`, `next` & `previous` are not. And I guess the user will realise pretty quick that the escape key doesn't close the dialog. I'd probably even leave out the "just this once" part as well.

frypf replied:
I used your css pretty-much as-is, although I switched the subtext colour for an opacity value in case Spotify invert their dialog colour scheme sometime in the future.

frypf replied:
Actually is it worth changing "Chrome" to be non-specific, as this extension should work with any chromium-based browser? I don't even use it with Chrome myself, other than for testing.

zopieux replied:
In believe this is resolved.

frypf commented on src/_locales/fr/messages.json:
```diff
@@ -5,6 +5,9 @@
         "message": "Ajoute des raccourcis clavier globaux pour le lecteur Spotify Web."
     },
+    "queue_search": {
+        "message": "Navigation: accueil / queue / rechercher"
```
Apologies for lazily using Google translate (applies to the next few comments as well 😅).

zopieux replied:
No worries you can't just magically be fluent. I'd suggest leaving it empty next time, and I'll add them.

frypf commented:
Funnily enough, it seemed pretty trivial when I first started, thinking _"Oh yeah, I could just add that one command, that seems easy enough"_. But as is so often the case, one issue leads to another, and pretty soon you've nearly doubled the codebase catering for a situation that no-one except yourself has ever even reported noticing 😂.

I'm a self-taught novice and my git-fu is decidedly weak, so I just wanted to clarify, when you say
> ...keeping the changes in their respective commits...

you're talking about leaving my original changes / commits intact and addressing the highlighted issues in new commits? I figure that's how it's normally done anyway, but as I said, spot the noob. Having read through, I'd envisage probably parcelling them up into another 3-4 commits depending on which bits make most sense to place together, rather than just doing a single "address-everything".

zopieux commented:
> you're talking about leaving my original changes / commits intact and addressing the highlighted issues in new commits

Hm no, precisely, the idea here would be to fix comments in the commit that is affected. I recon it's a lot of work though (you need to "rebase" etc.), so what you can do instead is [squash](https://fle.github.io/git-tip-keep-your-branch-clean-with-fixup-and-autosquash.html) all your commits into one, then work on addressing the review comments, then `commit --amend` that one commit, then `push --force-with-lease` to your branch (will update this PR). In other words I'd prefer fewer commits than 10 one-liners. Thanks.

frypf commented:
Ha, so exactly the opposite of what I said 🤦‍♂️. Thanks for the link, that's useful info.

zopieux commented:
Thanks. Will review soon.

zopieux commented on src/_locales/fr/messages.json:
```diff
@@ -37,5 +40,11 @@
+        "message": "Cliquez n'importe où ou appuyez sur une touche."
+    },
+    "interact_dialog_subtext": {
+        "message": "Cette interaction est une mesure de sécurité imposée par Chrome, après laquelle les raccourcis fonctionneront de manière transparente."
```
Chrome → le navigateur

zopieux commented on src/manifest.json:
```diff
@@ -10,10 +10,10 @@
         "48": "icon-48.png",
         "128": "icon-128.png"
     },
-    "incognito": "split",
```
What's wrong with split?

frypf replied:
Unfortunately without switching to the default "spanning", the extension runs completely separate service workers for incognito windows vs normal windows so any commands dispatched from an incognito window are unable to be sent to a spotify tab in another normal window.

frypf replied:
I just tested from within the incognito service worker devtools on the current 1.2.1 version and `await chrome.tabs.query({ url: 'https://open.spotify.com/*' })` always returns an empty array unless a spotify tab exists somewhere within an incognito window.

frypf replied:
Was this change ok? I think everything else is resolved, but I can't think of any other way around this one.

zopieux replied:
I don't have a strong opinion. Having this extension work standalone in incognito is not a high priority. If you need to remove `split` for this to work it's okay with me.

frypf replied:
Yeah I can't imagine that many people would be trying to run separate normal/incognito Spotify tabs at the same time, and if we just want everything to work seamlessly otherwise I'm inclined to leave it out.

If there are any privacy / security concerns, they already have to explicitly go in to the extension details page and allow in incognito anyway 👍.

zopieux commented:
Can you check the PR checkbox "Allow edits from maintainers" please? I would like to rebase and eventually make adjustments before merging.

frypf commented:
Yep, done. Sorry I prob should've checked this at the outset.

zopieux commented:
I think I found a undesired behavior.

1. Start from no Spotify tab at all, or from an existing tab
2. Use the home/search/queue shortcut, which should open a new Spotify tab at "home"
3. Use it again, which should navigate to "queue"

Notice the "interaction needed" dialog is displayed even though I haven't done anything that would require an interaction. It's also not vertically aligned, so I think it's not intended.

If you continue cycling, dialog stays visible for both "queue" and "home", but is hidden for "search".

I don't think this is intended. Can you try figuring this out please?

zopieux commented on src/manifest.json:
```diff
@@ -10,10 +10,10 @@
     "background": {
         "service_worker": "background.js"
     },
+    "action": { },
```
Is this empty object necessary?

frypf replied:
Oh that's just to make sure that the browser action is pinnable / clickable. Depending whether you're inclined to keep it, could switch it for

"action": {
"default_title": "Switch to / create Spotify tab."
},
for a little more explanation? (Would require an additional localisation message though).

Or just omit it entirely and focus solely on the keyboard shortcuts. I had originally wanted to use it for the 'queue-search' command until I realised chrome disallows you from setting the browser action as a global keyboard shortcut.

zopieux replied:
Nah it's okay, let's keep it as-is.

zopieux commented:
Ah fuck, I probably overwrote your last 2 commits. Let me fix that.

frypf commented:
If anything, it's reassuring that I'm not the only one who sometimes does unintended things while using git 😂.

frypf commented on src/web_accessible/checkPageInteraction.js:
```diff
@@ -0,0 +1,97 @@
+      delete dialog.dataset.swphLastCommand;
+      // Allow other commands but with unobtrusive visible warning, unless switching
+      // to search page - don't want to get in the way when the search input element
+      // receives focus automatically.
```
I've realised if you do retain the non-modal dialog for certain situations, we also need to add a high positive z-index to the `dialogStyle` css. It would catch a rare case where the dialog element is added prior to the main page-body being visible.
(If you open / reload the spotify tab and then _immediately_ attempt hotkey playback the modal is always shown correctly above everything, but when it switches to a normal dialog it's hidden below the `#main` div).

frypf replied:
Just depends which way you'd prefer. I don't know if it's potentially more confusing to have the dialog disappear entirely or to have it occasionally presented in a non-blocking manner after the user has seen but not acted upon the modal version.

zopieux replied:
If I understand correctly this discussion, I'd just drop the concept of "non modal / 'priming' dialog" and go with the simple approach of requiring attention (showing the modal) when absolutely necessary (playback stuff). Sounds good?

frypf replied:
Yep I've implemented option 1. (As we still need to delete the dataset prop to avoid unexpectedly dispatching play / next / previous after the user has issued a different command before interacting).

frypf reviewed (commented):
Realised these comments might not be showing up properly after the last force-push (they were stuck at "pending"). Apologies if you had already seen them.

frypf commented:
Yep, no idea what happened there, other than some github weirdness between me writing my responses you undoing your last push.

I'd originally made [this comment](https://github.com/zopieux/spotify-hotkeys/pull/15#discussion_r1306500529) about an hour ago after the other comments regarding the dialog, but for some reason it got duplicated but completely orphaned from your original question. I've deleted and re-added which seems to have got rid of the duplicate, but frustratingly it's out of order now, so I added the link ref to the dialog discussion to try and make it clearer.

zopieux commented on src/web_accessible/checkPageInteraction.js:
```diff
@@ -0,0 +1,97 @@
+ * Checks user interacted with page before dispatching `command`.
+ * - Presents modal dialog if attempting to dispatch `'play-pause'`/`'next'`/`'previous'`.
+ * - Presents non-blocking dialog for all other commands.
+ * - Mitigates effect of `DOMException: play() failed because the user didn't interact with the document first`.
```
>Presents non-blocking dialog for all other commands

Ah ok missed that apologies. I'm indeed not a big fan, it's too obnoxious for a "priming", to my taste. I'd just not display it until really required, if you don't mind making the change?

zopieux commented:
All right, tested the latest state, looks great! :100:

frypf commented:
Awesome, glad to be of assistance! Cheers for putting up with my git naivety.

zopieux commented:
Version 1.3.1 is under review by Google.

zopieux commented:
1.3.1 is out.

frypf commented:
Shall I go ahead and delete my fork?

zopieux commented:
You can, yes.
